README.md
# Numeric separator support for Acorn

[![NPM version](https://img.shields.io/npm/v/acorn-numeric-separator.svg)](https://www.npmjs.org/package/acorn-numeric-separator)

This is a plugin for [Acorn](http://marijnhaverbeke.nl/acorn/) - a tiny, fast JavaScript parser, written completely in JavaScript.

It implements support for numeric separators as defined in the stage 3 proposal [Numeric Separators](https://github.com/tc39/proposal-numeric-separator).

## Usage

This module provides a plugin that can be used to extend the Acorn `Parser` class to parse numeric separators.
You can either choose to use it via CommonJS (for example in Node.js) like this

```javascript
const {Parser} = require('acorn');
const numericSeparator = require('acorn-numeric-separator');
Parser.extend(numericSeparator).parse('100_000');
```

or as an ECMAScript module like this:

```javascript
import {Parser} from 'acorn';
import numericSeparator from 'path/to/acorn-numeric-separator.mjs';
Parser.extend(numericSeparator).parse('100_000');
```

## License

This plugin is released under an [MIT License](./LICENSE).


CHANGELOG.md
## 0.3.5 (2020-08-13)

* Actually include ESM files in release

## 0.3.3 (2020-06-13)

* Migrate to ESM

## 0.3.2 (2020-05-06)

* Also remove numeric separators in Literal's `bigint` property

## 0.3.1 (2020-05-03)

* Make compatible with acorn's bigint support
* Mark as compatible with acorn@7

## 0.3.0 (2019-04-04)

* Make compatible with acorn-bigint

## 0.2.0 (2018-09-14)

* Update to new acorn 6 interface
* Change license to MIT

## 0.1.1 (2018-01-16)

* Don't bail on empty integers as in `1.`

## 0.1.0 (2017-12-19)

Initial release


test/test.js
"use strict"

const assert = require("assert")
const acorn = require("acorn")
const numericSeparator = require("..")
const Parser = acorn.Parser.extend(numericSeparator)

function test(text, expectedResult, additionalOptions) {
  it(text, function () {
    const result = Parser.parse(text, Object.assign({ ecmaVersion: 9, plugins: { numericSeparator: true } }, additionalOptions))
    assert.deepStrictEqual(result.body[0], expectedResult)
  })
}
function testFail(text, expectedError, additionalOptions) {
  it(text, function () {
    let failed = false
    try {
      Parser.parse(text, Object.assign({ ecmaVersion: 9, plugins: { numericSeparator: true } }, additionalOptions))
    } catch (e) {
      assert.strictEqual(e.message, expectedError)
      failed = true
    }
    assert(failed)
  })
}

const newNode = (start, props) => Object.assign(new acorn.Node({options: {}}, start), props)

describe("acorn-numeric-separator", function () {
  testFail("'\\u{12_34}'", "Bad character escape sequence (1:4)")
  testFail("'\\u12_34'", "Bad character escape sequence (1:3)")
  testFail("let a\\u{12_34} = 5", "Bad character escape sequence (1:8)")

  const digits = [
    {d: "1_0", ast: start => newNode(start, {
      type: "Literal",
      end: start + 3,
      value: 10,
      raw: "1_0",
    })},
    {d: "12e3_4", ast: start => newNode(start, {
      type: "Literal",
      end: start + 6,
      value: 12e34,
      raw: "12e3_4",
    })},
    {d: "1_2e34", ast: start => newNode(start, {
      type: "Literal",
      end: start + 6,
      value: 12e34,
      raw: "1_2e34",
    })},
    {d: "0b1010_1001", ast: start => newNode(start, {
      type: "Literal",
      end: start + 11,
      value: 169,
      raw: "0b1010_1001",
    })},
    {d: "0xA0_B0_C0", ast: start => newNode(start, {
      type: "Literal",
      end: start + 10,
      value: 0xa0b0c0,
      raw: "0xA0_B0_C0",
    })},
    {d: "0o70_60_50", ast: start => newNode(start, {
      type: "Literal",
      end: start + 10,
      value: 0o706050,
      raw: "0o70_60_50",
    })},

    {d: "_2", ast: start => newNode(start, {
      type: "Identifier",
      end: start + 2,
      name: "_2",
    })},
    {d: "0b_1", error: start => `Invalid numeric separator (1:${start + 2})`},
    {d: "0o_1", error: start => `Invalid numeric separator (1:${start + 2})`},
    {d: "0x_1", error: start => `Invalid numeric separator (1:${start + 2})`},
    {d: "2_", error: start => `Invalid numeric separator (1:${start + 1})`},
    {d: "2__4", error: start => `Invalid numeric separator (1:${start + 2})`},
    {d: "._4", error: start => `Unexpected token (1:${start})`},
    {d: "_.4", error: start => `Unexpected token (1:${start + 1})`},
    {d: "1._4", error: start => `Invalid numeric separator (1:${start + 2})`},
    {d: "1_.4", error: start => `Invalid numeric separator (1:${start + 1})`},
    {d: "_1.4", error: start => `Unexpected token (1:${start + 2})`},
    {d: "1.4_", error: start => `Invalid numeric separator (1:${start + 3})`},
    {d: "1.4_e2", error: start => `Invalid numeric separator (1:${start + 3})`},
    {d: "1.4e_2", error: start => `Invalid numeric separator (1:${start + 4})`},
    {d: "04_3_2", error: start => `Invalid number (1:${start})`},
    {d: "0_4_3_2", error: start => `Invalid number (1:${start})`},
  ]
  const statements = [
    {s: "let i = %s", ast: content => newNode(0, {
      type: "VariableDeclaration",
      end: content.end,
      kind: "let",
      declarations: [newNode(4, {
        type: "VariableDeclarator",
        end: content.end,
        id: newNode(4, {
          type: "Identifier",
          end: 5,
          name: "i"
        }),
        init: content
      })]
    })},

    {s: "i = %s", ast: content => newNode(0, {
      type: "ExpressionStatement",
      end: content.end,
      expression: newNode(0, {
        type: "AssignmentExpression",
        end: content.end,
        operator: "=",
        left: newNode(0, {
          type: "Identifier",
          end: 1,
          name: "i"
        }),
        right: content
      })
    })},

    {s: "((i = %s) => {})", ast: content => newNode(0, {
      type: "ExpressionStatement",
      end: content.end + 8,
      expression: newNode(1, {
        type: "ArrowFunctionExpression",
        end: content.end + 7,
        id: null,
        generator: false,
        expression: false,
        async: false,
        params: [
          newNode(2, {
            type: "AssignmentPattern",
            end: content.end,
            left: newNode(2, {
              type: "Identifier",
              end: 3,
              name: "i"
            }),
            right: content
          })
        ],
        body: newNode(content.end + 5, {
          type: "BlockStatement",
          end: content.end + 7,
          body: []
        })
      })
    })},

    {s: "for (let i = 10; i < %s;++i) {}", ast: content => newNode(0, {
      type: "ForStatement",
      end: content.end + 8,
      init: newNode(5, {
        type: "VariableDeclaration",
        end: 15,
        declarations: [
          newNode(9, {
            type: "VariableDeclarator",
            end: 15,
            id: newNode(9, {
              type: "Identifier",
              end: 10,
              name: "i"
            }),
            init: newNode(13, {
              type: "Literal",
              end: 15,
              value: 10,
              raw: "10"
            })
          })
        ],
        kind: "let"
      }),
      test: newNode(17, {
        type: "BinaryExpression",
        end: content.end,
        left: newNode(17, {
          type: "Identifier",
          end: 18,
          name: "i"
        }),
        operator: "<",
        right: content
      }),
      update: newNode(content.end + 1, {
        type: "UpdateExpression",
        end: content.end + 4,
        operator: "++",
        prefix: true,
        argument: newNode(content.end + 3, {
          type: "Identifier",
          end: content.end + 4,
          name: "i"
        })
      }),
      body: newNode(content.end + 6, {
        type: "BlockStatement",
        end: content.end + 8,
        body: []
      })
    })},

    {s: "i + %s", ast: content => newNode(0, {
      type: "ExpressionStatement",
      end: content.end,
      expression: newNode(0, {
        type: "BinaryExpression",
        end: content.end,
        left: newNode(0, {
          type: "Identifier",
          end: 1,
          name: "i"
        }),
        operator: "+",
        right: content
      })
    })}
  ]
  statements.forEach(statement => {
    const start = statement.s.indexOf("%s")
    digits.forEach(d => {
      (d.error ? testFail : test)(
        statement.s.replace("%s", d.d),
        d.error ? d.error(start) : statement.ast(d.ast(start))
      )
    })
  })

  // Make sure we didn't break anything
  test("123..toString(10)", newNode(0, {
    type: "ExpressionStatement",
    end: 17,
    expression: newNode(0, {
      type: "CallExpression",
      end: 17,
      callee: newNode(0, {
        type: "MemberExpression",
        end: 13,
        object: newNode(0, {
          type: "Literal",
          end: 4,
          raw: "123.",
          value: 123
        }),
        property: newNode(5, {
          type: "Identifier",
          end: 13,
          name: "toString"
        }),
        computed: false,
      }),
      arguments: [
        newNode(14, {
          type: "Literal",
          start: 14,
          end: 16,
          raw: "10",
          value: 10
        })
      ],
    })
  }))
})


inject.js
"use strict"

module.exports = function (acorn) {
  acorn.plugins.numericSeparator = function (instance) {
    instance.extend("readInt", function (superF) {
      return function(radix, len) {
        // Hack: len is only != null for unicode escape sequences,
        // where numeric separators are not allowed
        if (len != null) return superF.call(this, radix, len)

        let start = this.pos, total = 0, acceptUnderscore = false
        for (;;) {
          let code = this.input.charCodeAt(this.pos), val
          if (code >= 97) val = code - 97 + 10 // a
          else if (code == 95) {
            if (!acceptUnderscore) this.raise(this.pos, "Invalid numeric separator")
            ++this.pos
            acceptUnderscore = false
            continue
          } else if (code >= 65) val = code - 65 + 10 // A
          else if (code >= 48 && code <= 57) val = code - 48 // 0-9
          else val = Infinity
          if (val >= radix) break
          ++this.pos
          total = total * radix + val
          acceptUnderscore = true
        }
        if (this.pos === start) return null
        if (!acceptUnderscore) this.raise(this.pos - 1, "Invalid numeric separator")

        return total
      }
    })

    instance.extend("readNumber", function (superF) {
      return function(startsWithDot) {
        const token = superF.call(this, startsWithDot)
        let octal = this.end - this.start >= 2 && this.input.charCodeAt(this.start) === 48
        const stripped = this.input.slice(this.start, this.end).replace(/_/g, "")
        if (stripped.length < this.end - this.start) {
          if (octal) this.raise(this.start, "Invalid number")
          this.value = parseFloat(stripped)
        }
        return token
      }
    })
  }

  return acorn
}


run_test262.js
"use strict"

const path = require("path")
const run = require("test262-parser-runner")
const acorn = require("acorn")
const Parser = acorn.Parser.extend(require("."))

const unsupportedFeatures = [
  "BigInt",
  "class-fields-private",
  "class-fields-public",
  "class-methods-private",
  "class-static-fields-private",
  "class-static-fields-public",
  "class-static-methods-private",
  "dynamic-import",
  "export-star-as-namespace-from-module",
  "import.meta"
]

run(
  (content, options) => Parser.parse(content, {sourceType: options.sourceType, ecmaVersion: 9}),
  {
    testsDirectory: path.dirname(require.resolve("test262/package.json")),
    skip: test => (!test.attrs.features || !test.attrs.features.includes("numeric-separator-literal") || unsupportedFeatures.some(f => test.attrs.features.includes(f))),
  }
)


index.js
"use strict"

function withoutAcornBigInt(acorn, Parser) {
  return class extends Parser {
    readInt(radix, len) {
      // Hack: len is only != null for unicode escape sequences,
      // where numeric separators are not allowed
      if (len != null) return super.readInt(radix, len)

      let start = this.pos, total = 0, acceptUnderscore = false
      for (;;) {
        let code = this.input.charCodeAt(this.pos), val
        if (code >= 97) val = code - 97 + 10 // a
        else if (code == 95) {
          if (!acceptUnderscore) this.raise(this.pos, "Invalid numeric separator")
          ++this.pos
          acceptUnderscore = false
          continue
        } else if (code >= 65) val = code - 65 + 10 // A
        else if (code >= 48 && code <= 57) val = code - 48 // 0-9
        else val = Infinity
        if (val >= radix) break
        ++this.pos
        total = total * radix + val
        acceptUnderscore = true
      }
      if (this.pos === start) return null
      if (!acceptUnderscore) this.raise(this.pos - 1, "Invalid numeric separator")

      return total
    }

    readNumber(startsWithDot) {
      const token = super.readNumber(startsWithDot)
      let octal = this.end - this.start >= 2 && this.input.charCodeAt(this.start) === 48
      const stripped = this.getNumberInput(this.start, this.end)
      if (stripped.length < this.end - this.start) {
        if (octal) this.raise(this.start, "Invalid number")
        this.value = parseFloat(stripped)
      }
      return token
    }

    // This is used by acorn-bigint
    getNumberInput(start, end) {
      return this.input.slice(start, end).replace(/_/g, "")
    }
  }
}

function withAcornBigInt(acorn, Parser) {
  return class extends Parser {
    readInt(radix, len) {
      // Hack: len is only != null for unicode escape sequences,
      // where numeric separators are not allowed
      if (len != null) return super.readInt(radix, len)

      let start = this.pos, total = 0, acceptUnderscore = false
      for (;;) {
        let code = this.input.charCodeAt(this.pos), val
        if (code >= 97) val = code - 97 + 10 // a
        else if (code == 95) {
          if (!acceptUnderscore) this.raise(this.pos, "Invalid numeric separator")
          ++this.pos
          acceptUnderscore = false
          continue
        } else if (code >= 65) val = code - 65 + 10 // A
        else if (code >= 48 && code <= 57) val = code - 48 // 0-9
        else val = Infinity
        if (val >= radix) break
        ++this.pos
        total = total * radix + val
        acceptUnderscore = true
      }
      if (this.pos === start) return null
      if (!acceptUnderscore) this.raise(this.pos - 1, "Invalid numeric separator")

      return total
    }

    readNumber(startsWithDot) {
      let start = this.pos
      if (!startsWithDot && this.readInt(10) === null) this.raise(start, "Invalid number")
      let octal = this.pos - start >= 2 && this.input.charCodeAt(start) === 48
      let octalLike = false
      if (octal && this.strict) this.raise(start, "Invalid number")
      let next = this.input.charCodeAt(this.pos)
      if (!octal && !startsWithDot && this.options.ecmaVersion >= 11 && next === 110) {
        let str = this.getNumberInput(start, this.pos)
        // eslint-disable-next-line node/no-unsupported-features/es-builtins
        let val = typeof BigInt !== "undefined" ? BigInt(str) : null
        ++this.pos
        if (acorn.isIdentifierStart(this.fullCharCodeAtPos())) this.raise(this.pos, "Identifier directly after number")
        return this.finishToken(acorn.tokTypes.num, val)
      }
      if (octal && /[89]/.test(this.input.slice(start, this.pos))) {
        octal = false
        octalLike = true
      }
      if (next === 46 && !octal) { // '.'
        ++this.pos
        this.readInt(10)
        next = this.input.charCodeAt(this.pos)
      }
      if ((next === 69 || next === 101) && !octal) { // 'eE'
        next = this.input.charCodeAt(++this.pos)
        if (next === 43 || next === 45) ++this.pos // '+-'
        if (this.readInt(10) === null) this.raise(start, "Invalid number")
      }
      if (acorn.isIdentifierStart(this.fullCharCodeAtPos())) this.raise(this.pos, "Identifier directly after number")
      let str = this.getNumberInput(start, this.pos)
      if ((octal || octalLike) && str.length < this.pos - start) {
        this.raise(start, "Invalid number")
      }

      let val = octal ? parseInt(str, 8) : parseFloat(str)
      return this.finishToken(acorn.tokTypes.num, val)
    }

    parseLiteral(value) {
      const ret = super.parseLiteral(value)
      if (ret.bigint) ret.bigint = ret.bigint.replace(/_/g, "")
      return ret
    }

    readRadixNumber(radix) {
      let start = this.pos
      this.pos += 2 // 0x
      let val = this.readInt(radix)
      if (val == null) { this.raise(this.start + 2, `Expected number in radix ${radix}`) }
      if (this.options.ecmaVersion >= 11 && this.input.charCodeAt(this.pos) === 110) {
        let str = this.getNumberInput(start, this.pos)
        // eslint-disable-next-line node/no-unsupported-features/es-builtins
        val = typeof BigInt !== "undefined" ? BigInt(str) : null
        ++this.pos
      } else if (acorn.isIdentifierStart(this.fullCharCodeAtPos())) { this.raise(this.pos, "Identifier directly after number") }
      return this.finishToken(acorn.tokTypes.num, val)
    }

    // This is used by acorn-bigint, which theoretically could be used with acorn@6.2 || acorn@7
    getNumberInput(start, end) {
      return this.input.slice(start, end).replace(/_/g, "")
    }
  }
}

module.exports = function(Parser) {
  const acorn = Parser.acorn || require("acorn")
  const withAcornBigIntSupport = (acorn.version.startsWith("6.") && !(acorn.version.startsWith("6.0.") || acorn.version.startsWith("6.1."))) || acorn.version.startsWith("7.")

  return withAcornBigIntSupport ? withAcornBigInt(acorn, Parser) : withoutAcornBigInt(acorn, Parser)
}
